Unix Sockets for HTTP Replication (#15708)
matrix-org/synapse: docs/usage/configuration/config_documentation.md

* `additional_resources`: Only valid for an 'http' listener. A map of
   additional endpoints which should be loaded via dynamic modules.

> UNIX socket support(_Added in Synapse 1.87.0_):

Reviewer: ```suggestion
Unix socket support (_Added in Synapse 1.87.0_):
```

> Unix socket support (_Added in Synapse 1.87.0_):

Author: (Note to self to update version number before merge)

Reviewer: I think this PR is ready to go (make this happen)

Author: Updated in [92c6c7b](https://github.com/matrix-org/synapse/pull/15708/commits/92c6c7b31de24e0174d12df241bd6bf713eb5786)

  * `path`: A path and filename for a Unix socket. Make sure it is located in a
    directory with read and write permissions, and that it already exists (the directory

>     will not be created). Defaults to `None`."

Reviewer: ```suggestion
    will not be created). Defaults to `None`.
```

    * **Note**: The use of both `path` and `port` options for the same `listener` is not
      compatible.
    * The `x_forwarded` option defaults to true  when using Unix sockets and can be omitted.
    * Other options that would not make sense to use with a UNIX socket, such as 
      `bind_addresses` and `tls` will be ignored and can be removed.
  * `mode`: The file permissions to set on the UNIX socket. Defaults to `666`

>   * **Note:** Must be set as `type: http` (does not support `metrics` and `manhole`). 

Reviewer: Doesn't affect the rendered result but it looks like all of these need to be un-indented to match the root list level of the surrounding things. Wish we had some auto-formatting for this.

Author: (forgot to link that) Fixed in [85eba85](https://github.com/matrix-org/synapse/pull/15708/commits/85eba85a4eb09496ccf7a2c77848f06e0cfe0178)

* `additional_resources`: Only valid for an 'http' listener. A map of
   additional endpoints which should be loaded via dynamic modules.

>   _Added in Synapse 1.86.0_

Reviewer: Currently, this gets grouped to the list item above. It should be added below the item or perhaps in parenthesis next to `Unix socket support (...):`

Author: The formatting leaves much to be desired. Look right above this section and you'll see a subsection(ish) that is for 'Added in Synapse 1.68.0' that blends in too. There are a bunch of spots where you can find various 'Added in Synapse x.xx.x' that look like this

Author: Actually, I wonder if I stick 4 `#`'s in front of it how it would look(Pycharm's render mode is only slightly helpful). Let's try it.

Author: https://pr15708--synapse-docs-previews.netlify.app/usage/configuration/config_documentation.html#listeners

Well, that's a little bolder I guess. What do you think?

Reviewer: The problem is that it's grouped in the `<li>` above where it doesn't apply.

The look could be improved but it aligns with how we do it for the rest of the things there so it's fine.

Author: Oh. I thought it was ok like this as it was a 'feature' of the `listeners` config section.

Author: I've thought about the potential of issues with using 4 `#`'s and decided it's not worth that, so I'll abandon that line of thought. Let's continue this with the other comment thread, as it's more relevant there.

* UNIX socket support:
  * `path`: A path and filename for a UNIX socket. Ensure it is in a readable/writeable
     directory. No default
    * **Note**: Declaring a `path` is not compatible while also having a `port`
       configuration for the same `listener`
    * **Note**: `metrics` and `manhole` are not supported as `types` on UNIX socket `listeners`
    * The `x_forwarded` option will default to true and may be left out.
    * Other options that would not make sense to use with a UNIX socket, such as 

>     `bind_address[es]` and `tls` will be ignored and can be removed.

Reviewer: Seems like we might just want to validate that these options aren't set to avoid people being confused about what they need to have configured. And to avoid some misguided reports like "I have `bind_address` and `tls` set but it doesn't seem to be applied"

Author: In the near future, I'm hoping to add in some basic 'fall-over' support. In those instances, leaving these in means that if a Unix socket file isn't found where it says it should be, it will start using the Tcp address and related configuration. That is only in the idea phase still, and I need some Opinions on it before I try and move that forward.

That being said, there is some basic validity checking in place for if a `path` and a `port` are declared simultaneously, [here](https://github.com/matrix-org/synapse/blob/99c850f79821e12ad1895b9505f8612752deea52/synapse/config/server.py#L948-L957). In retrospect, these may not be fully sufficient. I will ponder on this, that is a fair concern to raise.

Author: Would you like this touched in *this* PR? Or should that be separate as 'scope creep'?

Reviewer: If we're not doing anything like that now and don't have easy utilities to add these constraints, we can push to a future iteration :+1:

>   * `mode`: The file permissions to set on the UNIX socket. Defaults to `666`

Reviewer: Why does the file `mode` matter to customize here? Seems like we should just use a sane default.

What's the least restrictive permissions we can use here?

Author: It felt to me that some who value the most locked down version of everything would appreciate being able to customize this. `666` is probably be the best for a default. However,(and I use Complement as my context) I believe anything more restrictive prohibits usage by nginx for the reverse proxy as it is a different 'user'(and would therefore classify as 'Everyone/No-one' which is what the last digit controls). As a visual test of what happens if I change the default to a more restrictive `660`:
```log
nginx | 2023/06/15 20:37:50 [crit] 74#74: *171 connect() to unix:/run/main_public.sock failed (13: Permission denied) while connecting to upstream, client: 172.17.5.1, server: localhost, request: "GET /_matrix/client/versions HTTP/1.1", upstream: "http://unix:/run/main_public.sock:/_matrix/client/versions", host: "127.0.0.1:32788"
```
floods the logs

In theory, you could set up a special user for both your reverse proxy and Synapse so they would share permissions. But that is an exercise for a security professional.

TLDR: Like a TCP port, you want anything and everything to be able to read/write it, authentication should be handled by the individual process on either side.

Reviewer: It seems like `666` is the sane default :+1: 

And the only other case really is `660` if they share the same user.

Mostly asking this question to avoid extra unnecessary config. I feel like we could get away without the config and just use `666`. People could always pre-create the socket file with their custom permissions.

Author: Unfortunately, pre-creating a socket file doesn't work(with or without custom permissions).
<details><summary>The logs when that happens</summary>

```log
synapse_main | 2023-06-16 21:44:48,715 - synapse.app._base - 269 - CRITICAL - sentinel - Error during startup
synapse_main | Traceback (most recent call last):
synapse_main |   File "/usr/local/lib/python3.11/site-packages/twisted/internet/unix.py", line 411, in startListening
synapse_main |     skt.bind(self.port)
synapse_main | OSError: [Errno 98] Address already in use
synapse_main |
synapse_main | During handling of the above exception, another exception occurred:
synapse_main |
synapse_main | Traceback (most recent call last):
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/_base.py", line 254, in wrapper
synapse_main |     await cb(*args, **kwargs)
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/homeserver.py", line 363, in start
synapse_main |     await _base.start(hs)
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/_base.py", line 588, in start
synapse_main |     hs.start_listening()
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/homeserver.py", line 251, in start_listening
synapse_main |     self._listener_http(self.config, listener)
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/homeserver.py", line 141, in _listener_http
synapse_main |     ports = listen_http(
synapse_main |             ^^^^^^^^^^^^
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/_base.py", line 435, in listen_http
synapse_main |     ports = listen_unix(
synapse_main |             ^^^^^^^^^^^^
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/_base.py", line 384, in listen_unix
synapse_main |     cast(Port, reactor.listenUNIX(path, factory, backlog, mode, wantPID))
synapse_main |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
synapse_main |   File "/usr/local/lib/python3.11/site-packages/twisted/internet/posixbase.py", line 270, in listenUNIX
synapse_main |     p.startListening()
synapse_main |   File "/usr/local/lib/python3.11/site-packages/twisted/internet/unix.py", line 413, in startListening
synapse_main |     raise error.CannotListenError(None, self.port, le)
synapse_main | twisted.internet.error.CannotListenError: Couldn't listen on any:b'/run/main_public.sock': [Errno 98] Address already in use.
synapse_main | Error during startup:
synapse_main | Traceback (most recent call last):
synapse_main |   File "/usr/local/lib/python3.11/site-packages/twisted/internet/unix.py", line 411, in startListening
synapse_main |     skt.bind(self.port)
synapse_main | OSError: [Errno 98] Address already in use
synapse_main |
synapse_main | During handling of the above exception, another exception occurred:
synapse_main |
synapse_main | Traceback (most recent call last):
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/_base.py", line 254, in wrapper
synapse_main |     await cb(*args, **kwargs)
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/homeserver.py", line 363, in start
synapse_main |     await _base.start(hs)
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/_base.py", line 588, in start
synapse_main |     hs.start_listening()
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/homeserver.py", line 251, in start_listening
synapse_main |     self._listener_http(self.config, listener)
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/homeserver.py", line 141, in _listener_http
synapse_main |     ports = listen_http(
synapse_main |             ^^^^^^^^^^^^
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/_base.py", line 435, in listen_http
synapse_main |     ports = listen_unix(
synapse_main |             ^^^^^^^^^^^^
synapse_main |   File "/usr/local/lib/python3.11/site-packages/synapse/app/_base.py", line 384, in listen_unix
synapse_main |     cast(Port, reactor.listenUNIX(path, factory, backlog, mode, wantPID))
synapse_main |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
synapse_main |   File "/usr/local/lib/python3.11/site-packages/twisted/internet/posixbase.py", line 270, in listenUNIX
synapse_main |     p.startListening()
synapse_main |   File "/usr/local/lib/python3.11/site-packages/twisted/internet/unix.py", line 413, in startListening
synapse_main |     raise error.CannotListenError(None, self.port, le)
synapse_main | twisted.internet.error.CannotListenError: Couldn't listen on any:b'/run/main_public.sock': [Errno 98] Address already in use.
```

</details>

Reviewer: Sounds like there is some extra locking or detail of the socket file content itself. Perhaps some command like this is necessary to create one: https://serverfault.com/a/914572/182342

Feels like it's possible to pre-create a socket file since the socket file stays behind after the app quits and starts again where it does work.

Reviewer: I run into the same error when using that command mentioned:

```
$ python -c "import socket as s; sock = s.socket(s.AF_UNIX); sock.bind('/tmp/test.sock')"
$ python -c "import socket as s; sock = s.socket(s.AF_UNIX); sock.bind('/tmp/test.sock')"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
OSError: [Errno 98] Address already in use
```

Not sure how the socket file normally gets cleaned up, etc so all of this works. But I guess `mode` is useful :shrug:

Author: The lock file contains the PID of the process that started it...it shouldn't stick around after the process exits, just like the socket file. I believe Twisted cleans it up automatically. It feels like we're branching a little into the theoretical with this...but yes it's possible to create the socket file externally, however that isn't behavior we should support. We want the server to create and clean up it's own socket like it does now. Creating an socket externally would require adding extra binding and other such stuff which complicates the matter a great deal.

That page you linked, the top answer defines the behavior we want(I think).

Reviewer: > It feels like we're branching a little into the theoretical with this...

I think it's mainly because I'm not familiar and just working on recall with the few times I've used them in other systems.

Using standard patterns makes sense and I think I have a better understanding now :+1:

Reviewer: And TLDR is `mode` is fine :fast_forward:

> * UNIX socket support:
>   * `path`: A path and filename for a UNIX socket. Ensure it is in a readable/writeable
>      directory and is pre-existing(the directory will not be created). No default
>     * **Note**: Declaring a `path` is not compatible while also having a `port`
>        configuration for the same `listener`
>     * **Note**: `metrics` and `manhole` are not supported as `types` on UNIX socket `listeners`
>     * The `x_forwarded` option will default to true and may be left out.
>     * Other options that would not make sense to use with a UNIX socket, such as 
>     `bind_address[es]` and `tls` will be ignored and can be removed.
>   * `mode`: The file permissions to set on the UNIX socket. Defaults to `666`
>   * `type: http` is valid (provided `metrics` is not included in the `resources`:`names`)

Reviewer: ```suggestion
UNIX socket support (_Added in Synapse 1.86.0_):

* `path`: A path and filename for a Unix socket. Make sure it is located in a directory with read and write permissions, and that it already exists (the directory will not be created). Defaults to `None`."
   * **Note**: Declaring a `path` is not compatible while also having a `port`
       configuration for the same `listener`
   * The `x_forwarded` option defaults to true  when using Unix sockets and can be omitted.
   * Other options that would not make sense to use with a UNIX socket, such as 
    `bind_addresses` and `tls` will be ignored and can be removed.
* `mode`: The file permissions to set on the Unix socket (must be at least TODO). Defaults to `666`
* **Note:** Must be set as `type: http` (does not support `metrics` and `manhole`). Also make sure that `metrics` is not included in `resources` -> `names`
```

Author: Some of that doesn't feel right...But I like most of it(Me bad at words ☹️).

The sub-heading of `Added in Synapse 1.86.0`, I would like to keep that just above to maintain consistency with the rest of the document(which I think was mentioned in another PR, I can find if you are interested). However, I do believe a good bolding would be nice(I tried it out in another comment chain [here](https://github.com/matrix-org/synapse/pull/15708#discussion_r1231571110)). Also need to change that to 1.87.0 now(and anywhere else it's mentioned).

`path` doesn't default to `None`, it has no default. `None` is an explicit value that is empty of meaning, so that's not right.

Remove the `(must be at least TODO)` from the `mode` line. (That was settled, right?)

The indentation changed a bit, was that intentional?

Mind if I tweak it using your suggestion but with those edits?

Reviewer: > `path` doesn't default to `None`, it has no default. `None` is an explicit value that is empty of meaning, so that's not right.

Looks like it defaults to `None`:

https://github.com/matrix-org/synapse/blob/2ac6c3bbb535677bd62b3df425dd1755dba79b66/synapse/config/server.py#L942

> Remove the `(must be at least TODO)` from the `mode` line. (That was settled, right?)

We can remove the `mode` stuff if the option is being removed

> The indentation changed a bit, was that intentional?

Yes as it didn't make sense to me to group "UNIX socket support" as another list item. And can just be more details like how "Valid resource names are:" is.

> Mind if I tweak it using your suggestion but with those edits?

Up to you if you think that's better.

Author: I really don't want to remove `mode` as a configurable. I almost included the only other option I could find(it was `wantPID` which controls if a lock file is created with the PID). I defaulted that to `True` as some light reading suggested it would make `systemd` and such systems happy, and was harmless for docker images/containers.

Author: Perhaps if it's documented in such a way as to be explicit it's for special circumstances with use by a security professional, and therefore just omit it for a standard installation?

Author: > Looks like it defaults to `None`:
> 
> https://github.com/matrix-org/synapse/blob/2ac6c3bbb535677bd62b3df425dd1755dba79b66/synapse/config/server.py#L942

Yeah, but look at the rest:
https://github.com/matrix-org/synapse/blob/0f02f0b4da92229e88e27a92ea3bfa523457bfc1/synapse/config/server.py#L942-L958

None is not allowed. Gosh this looks messy, in retrospect.

> > The indentation changed a bit, was that intentional?
> 
> Yes as it didn't make sense to me to group "UNIX socket support" as another list item. And can just be more details like how "Valid resource names are:" is.

I think I see what you mean by this, after staring at it for a while. Let me see what I can do about that with everything else.

Reviewer: > None is not allowed. Gosh this looks messy, in retrospect.

I don't see anything there that doesn't allow `None`. You just have to have `port` or `socket_path`.

In any case, this is really about how we want to explain defaults like this. Just make a decision and we can further nit from there if needed.

Author: It's simple enough. We say there is no default. Programmatically, yes it will default to `None` which means it would have to be non-existent in the yaml, or left empty. It's considered a 'special' value in this context. However, if it *is* `None` and there is no `port` defined as an alternative, then the `ConfigError` raises telling them they need to pick one. We don't choose a default for this(not only because every system is different) as it would fall to that automatically and maybe they just misconfigured a `port`(which incidentally also doesn't really have a default either)

Author: Updated in [39597fc](https://github.com/matrix-org/synapse/pull/15708/commits/39597fcf0bb13d41964cf400a5b6d06992eec926)

(except this most current part about no defaults)

* `media`: the media API (/_matrix/media).

> * `metrics`: the metrics interface. See [here](../../metrics-howto.md). (Not usable on a UNIX Socket `listener`)

Reviewer: > (Not usable on a UNIX Socket `listener`)

Why?

Author: The short story is that it's not servable in it's present form. The long story is in the other comment https://github.com/matrix-org/synapse/pull/15708#discussion_r1227922083

Reviewer: ```suggestion
* `metrics`: the metrics interface. See [here](../../metrics-howto.md). (Not compatible with Unix sockets)
```

Reviewer: Can we throw a config error if someone uses a Unix socket with `metrics`?

    bind_addresses: ['::1', '127.0.0.1']
    type: manhole
```
Example configuration #3:
```yaml
listeners:

>   # Unix socket listener: Ideal for Synapse deployments behind a reverse proxy, offering lightweight interprocess communication without TCP/IP overhead, avoid port conflicts, and providing enhanced security through system file permissions.

Reviewer: (wrap line)

Author: Fixed in [2f20dda](https://github.com/matrix-org/synapse/pull/15708/commits/2f20ddae4991ce5521822baefa5bb8630c9d220f)

    bind_addresses: ['::1', '127.0.0.1']
    type: manhole
```
Example configuration #3:
```yaml
listeners:

>   # UNIX socket listener: for when Synapse is behind a reverse proxy that can utilise such.

Reviewer: ```suggestion
  # Unix socket listener: for when Synapse is behind a reverse proxy and (TODO: better reason).
```

Author: I don't think there is a better reason. It will only work if there is a way to proxy incoming requests to a socket file. I think my American vernicular was sticking it's nose in here. Probably could just remove the `that can utilise such`.

Author: I think they were mentioned in the original issue:
> It would be useful for people who don't want to open a localhost TCP port when they have a proxy in front of synapse.

> +1 to this request. It's really common practice for app servers like gunicorn, to skip local networking and just use sockets between the frontend (like nginx) and the app server, e.g. http://docs.gunicorn.org/en/stable/deploy.html

> This would be cool. I am not really a fan of using unprivileged TCP ports for loopbacks. Any unprivileged local user could seize the port if the server shuts down for a while. This issue could perhaps therefore be tagged with the "security" tag?

Author: Furthermore, as I understand it, there is slightly less overhead with Unix socket versus TCP sockets. A lot of the baggage that was built into TCP/IP over the decades to handle a bunch of issues(bufferbloat, blackholes, misconfigured anything, etc) that had to do with connectivity, Unix sockets just....don't have.

Reviewer: Thanks for the context! Perhaps this:

```suggestion
  # Unix socket listener: Ideal for Synapse deployments behind a reverse proxy, offering lightweight interprocess communication without TCP/IP overhead, avoid port conflicts, and providing enhanced security through system file permissions.
```

  #
  # Note that x_forwarded will default to true, when using a UNIX socket. Please see
  # https://matrix-org.github.io/synapse/latest/reverse_proxy.html.
  #
  - path: /var/run/synapse/main_public.sock

>     mode: 660

Reviewer: Based off the [other comment](https://github.com/matrix-org/synapse/pull/15708#discussion_r1231518835), it seems like this would give permission errors to people. We should set a more sane default for the example here.

Author: That's fair. I can take it off, as less-is-more probably would apply here.

Author: Fixed in [9882e0a](https://github.com/matrix-org/synapse/pull/15708/commits/9882e0afdc00cb92f70786ad3807598e230fe93d)

>   - path: /var/run/synapse/main_public.sock
>     mode: 660
>     type: http

Author: ```suggestion
  - path: /var/run/synapse/main_public.sock
    type: http
```

    resources:

>       - names: [client, federation]

Reviewer: Does it make sense to use unix sockets for anything other than `replication`?

Author: > Why would someone want to put `client` or `federation` over unix sockets? Are you able to have an nginx layer in front of Synapse proxy through from the unix socket?

Yes, that's it exactly. I forget you don't have the benefit of doing the other PRs for context. When using a Unix socket, you ***have*** to put a reverse proxy of some sort in front of it. Unix sockets are not capable of being served(on their own) over the interwebs. Most reverse proxies(like nginx) are the layer that handles this for you(as well as things like TLS and such).
> 
> Thanks for the background on `metrics`! Makes sense since it's served by something separate.

You're very welcome!
